src/NosCore.GameObject/Services/ExchangeAccess/ExchangeAccessService.cs

```
@@ -0,0 +1,217 @@
+            }
+            
+            ResetExchangeData(session);
+            session.SendPacket(new ExcClosePacket { Type = closeType });
+        }
+
+        public void ResetExchangeData(ClientSession session)
+        {
+            if (session == null)
+            {
+                return;
+            }
+
+            session.Character.InExchange = false;
+            session.Character.ExchangeData = new ExchangeData();
+            session.Character.ExchangeRequests = new ConcurrentDictionary<Guid, long>();
+        }
+
+        public void ProcessExchange(ClientSession session, ClientSession targetSession)
+        {
+            foreach (var item in session.Character.ExchangeData.ExchangeItems.Values)
+            {
+                if (session.Character.Inventory.LoadByItemInstanceId<IItemInstance>(item.Id)?.Amount >= item.Amount)
+                {
+                    session.Character.Inventory.RemoveItemAmountFromInventory(item.Amount, item.Id);
+                    var temp2 = session.Character.Inventory.LoadBySlotAndType<IItemInstance>(item.Slot, item.Type);
+
+                    session.SendPacket(temp2.GeneratePocketChange(item.Type, item.Slot));
+                }
+                else
+                {
+                    _logger.Error(Language.Instance.GetMessageFromKey(LanguageKey.NOT_ENOUGH_ITEMS, session.Account.Language));
+                    return;
+                }
+            }
+            
+            foreach (var item in session.Character.ExchangeData.ExchangeItems.Values)
+            {
+                var itemCpy = _itemBuilderService.Create(item.ItemVNum, targetSession.Character.CharacterId,
+                    amount: item.Amount, rare: (sbyte) item.Rare, upgrade: item.Upgrade, design: (byte) item.Design);
+
+                var inv = targetSession.Character.Inventory.AddItemToPocket(itemCpy).FirstOrDefault();
+
+                if (inv == null)
+                {
+                    continue;
```

should throw an error